framework/src/modules/chain/bft/consensus_manager.js

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const assert = require('assert');
+const debug = require('debug')('lisk:bft:consensus_manager');
+const { HeadersList } = require('./headers_list');
+const { validateBlockHeader } = require('./utils');
+const {
+	BFTChainDisjointError,
+	BFTLowerChainBranchError,
+	BFTForkChoiceRuleError,
+	BFTInvalidAttributeError,
+} = require('./errors');
+
+/**
+ * @typedef {Object} BlockHeader
+ * @property {string} blockId
+ * @property {int} height
+ * @property {int} maxHeightPreviouslyForged
+ * @property {int} prevotedConfirmedUptoHeight
+ * @property {int} activeSinceRound
+ * @property {string} delegatePublicKey
+ */
+
+class ConsensusManager {
+	constructor({ finalizedHeight, activeDelegates } = {}) {
+		assert(finalizedHeight !== undefined, 'Must provide finalizedHeight');
+		assert(activeDelegates !== undefined, 'Must provide activeDelegates');
+		assert(activeDelegates > 0, 'Must provide a positive activeDelegates');
+
+		// Set constants
+		this.activeDelegates = activeDelegates;
+
+		// Threshold to consider a block pre-voted
+		this.preVoteThreshold = Math.ceil((this.activeDelegates * 2) / 3);
+
+		// Threshold to consider a block pre-committed (or finalized)
+		this.preCommitThreshold = Math.ceil((this.activeDelegates * 2) / 3);
+
+		// Limit to of blocks to must have to make any prevote-precommit verification on blocks (1 block less than 3 rounds)
+		this.processingThreshold = this.activeDelegates * 3 - 1;
+
+		// Maximum headers to store (5 rounds)
+		this.maxHeaders = this.activeDelegates * 5;
+
+		this.headers = new HeadersList({ size: this.maxHeaders });
+
+		// Height up to which blocks are finalized
+		this.finalizedHeight = finalizedHeight || 0;
```

Comment:
Maybe we could set this in the constructor `{ finalizedHeight = 0, activeDelegates } = {}`